# Sipros Ensemble Output

Sipros is a database-searching algorithm for peptide and protein identification in shotgun meta/proteomics. The output format of the various final and intermediate files generated by the Sipros are described below.

## Final Result Files

1. ${output_dir}/${ms2_filename}_${search_name}.tab : 

2. ${output_dir}/${ms2_filename}_${search_name}.psm.txt :

```
#	Filename = Filename of input FT2 file
#	ScanNumber = Scan number of the PSM
#	ParentCharge = Charge state of the PSM
#	MeasuredParentMass = Measured parent mass
#	CalculatedParentMass = Calculated parent mass from peptide sequence
#	MassErrorDa = Mass error in Da with 1-Da error correction
#	MassErrorPPM = Mass error in PPM with 1-Da error correction
#	ScanType = Scan type of the PSM
#	SearchName = Sipros search name
#	ScoringFunction = Scoring function used in the search
#	Score = Score
#	DeltaZ = Difference between the best PSM score and the next best PSM of this scan
#	DeltaP = Difference between the best modified PSM and its PTM isoform
#	IdentifiedPeptide = Identified peptide sequence with potential PTMs and mutations
#	OriginalPeptide = Original peptide sequence in the FASTA file
#	ProteinNames = Names of proteins of the peptide
#	ProteinCount = Number of proteins that the peptide can be assigned to
#	TargetMatch = T for target match and F for decoy match
```

3. ${output_dir}/${ms2_filename}_${search_name}.pep.txt :

4. ${output_dir}/${ms2_filename}_${search_name}.pro.txt :

5. ${output_dir}/${ms2_filename}_${search_name}.pro2pep.txt :

6. ${output_dir}/${ms2_filename}_${search_name}.pro2psm.txt :